src/rules/no-cycle.js

```diff
@@ -52,11 +62,36 @@ module.exports = {
 
     const options = context.options[0] || {};
     const maxDepth = typeof options.maxDepth === 'number' ? options.maxDepth : Infinity;
-    const ignoreModule = (name) => options.ignoreExternal && isExternalModule(
-      name,
-      resolve(name, context),
-      context,
-    );
+    
+    const ignoreModule = (name) => {
+      // Note: This fn is optimized to eliminate call time because it may be
+      // called hundreds of thousands of times during a full repository lint run
+      // and can add minutes to the total time in large codebases
+
+      // Low cost evaluation; do not ignore if not opted in or module starts
+      // with a relative path
+      // NOTE: skipping further resolve codepath also ensures that only unique
+      // import paths are set in resolvedPaths, eg "./index.js", or "../file.js"
+      // should not be stored because there could be duplicates that resolve to
+      // different paths
+      if (!options.ignoreExternal || name[0] === '.') return false;
```

Comment:
couldn't a resolver choose to configure a relative path as external?

Reply:
🤔 I suppose so, there's probably some monorepo setup out there like that...

I think removing this assumption means that we can't confidently cached the resolved module path we looked up because then there could be duplicate import paths that resolve to different files (eg import paths of `'./index.js'`).

Trying to approach this from a different angle: The current time to determine if a module is "external" is the performance bottleneck and this PR is attempting to short circuit that whenever possible. 

Do you have suggestions on an approach?

Reply:
What if the no-cycle rule option `ignoreExternal` could be a boolean or a regex, and if it's a regex it's used for this short circuit?

Reply:
I'm not sure what the best configuration fields are, but doing a quick test with a regex short circuit has about the same performance improvement as the cache implementation.

Alternative approach tested:
```js
    const ignoreModule = (name) => {
      if (!options.ignoreExternal ) return false;

      if (options.ignoreExternalRegex) {
        return options.ignoreExternalRegex.test(name);
      }

      // High cost evaluation; cache evaluations of resolve() and
      // isExternalModule() calls for faster subsequent lookups
      const path = resolve(name, context);

      let isExternal = externalModules[path];
      if (isExternal === undefined) {
        isExternal = isExternalModule(name, path, context);
        externalModules[path] = isExternal;
      }
      
      return isExternal;
    };
```

Reply:
Maybe the better question is, why is `isExternalModule` slow?

Perhaps the node and webpack resolvers can add a cache there, rather than inside the main plugin?

Reply:
Aren't we doing a bunch of regex matching in `isExternalModule`?

Reply:
generally that's fast, but if it can be sped up, let's!